server/src/main/java/org/apache/iotdb/db/engine/compaction/CompactionTaskManager.java

```diff
@@ -290,35 +292,34 @@ public synchronized void removeRunningTaskFromList(AbstractCompactionTask task)
    *
    * @throws RejectedExecutionException
    */
-  public synchronized void submitTask(Callable<Void> compactionMergeTask)
+  public synchronized Future<Void> submitTask(AbstractCompactionTask compactionTask)
       throws RejectedExecutionException {
     if (taskExecutionPool != null && !taskExecutionPool.isTerminated()) {
-      taskExecutionPool.submit(compactionMergeTask);
-      return;
+      Future<Void> future = taskExecutionPool.submit(compactionTask);
+      runningCompactionTaskList.add(compactionTask);
+      futureList.add(new Pair<>(compactionTask, future));
+      return future;
     }
     logger.warn(
         "A CompactionTask failed to be submitted to CompactionTaskManager because {}",
         taskExecutionPool == null
             ? "taskExecutionPool is null"
             : "taskExecutionPool is terminated");
+    return null;
   }
 
   /**
    * Abort all compactions of a storage group. The caller must acquire the write lock of the
    * corresponding storage group.
    */
-  public void abortCompaction(String fullStorageGroupName) {
-    Set<Future<Void>> subTasks =
-        storageGroupTasks.getOrDefault(fullStorageGroupName, Collections.emptySet());
-    candidateCompactionTaskQueue.clear();
-    Iterator<Future<Void>> subIterator = subTasks.iterator();
-    while (subIterator.hasNext()) {
-      Future<Void> next = subIterator.next();
-      if (!next.isDone() && !next.isCancelled()) {
-        next.cancel(true);
+  public synchronized void abortCompaction(String storageGroupName) {
+    for (Pair<AbstractCompactionTask, Future<Void>> futurePair : futureList) {
+      if (futurePair.left.getFullStorageGroupName().contains(storageGroupName)) {
+        futurePair.right.cancel(true);
       }
-      subIterator.remove();
     }
+
+    candidateCompactionTaskQueue.clear();
```

Comment:
Is it necessary to remove the task from` futureList` and `runningCompactionTaskList` here?

Reply:
When the compaction task finish, it will remove from the `futureList` and `runningCompactionTaskList` itself

Reply:
Got it.